### PR TITLE
*: local read when in leader's lease

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Implemented local read when in leader's lease with check_quorum is on.

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
@siddontang @ngaut @disksing @BusyJay PTAL

**Comment by siddontang** [comment]
Don't use config, we must support both lease read and linearizable read at same time. 
we have a `read_quorum` flag in command, we need this to determine whether using lease read or not. 
Like etcd, the default read is not linearizability, but user can set `read_quorum = true` explicitly.

**Comment by siddontang** [comment]
Now the default read is lease read.

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -771,9 +771,18 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         // log entry can't be committed.
 
 
-        // TODO: support handing read-only commands later.
         // for read-only, if we don't care stale read, we can
         // execute these commands immediately in leader.
+        if self.cfg.local_read_in_lease &&
+           peer.raft_group.raft.in_lease(self.cfg.leader_lease_safe_distance) &&
+           msg.get_requests().len() == 1 &&
+           msg.get_requests()[0].get_cmd_type() == CmdType::Snap {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: can we meet multiply requests all are snap?

File: src/raft/raft.rs [hunk 2]
```diff
@@ -304,6 +304,11 @@ impl<T: Storage> Raft<T> {
         hs
     }
 
+    pub fn in_lease(&self, safe_distance: usize) -> bool {
+        self.state == StateRole::Leader && self.check_quorum &&
+        self.election_elapsed + safe_distance <= self.election_timeout
```
Comment (siddontang) [inline src/raft/raft.rs]: the election_elapsed can only be reset after reach election_timeout, assume election_timeout is 20 ticks, so if the leader is already leader, we may still use raft read in [15, 20] ticks? 
only using check_quorum and state check is enough ? /cc @swingbach

File: src/raftstore/store/store.rs (hunk 1 above)
Comment (BusyJay) [inline src/raftstore/store/store.rs]: I think we should keep this logic in peer.rs

File: src/raft/raft.rs (hunk 2 above)
Comment (BusyJay) [inline src/raft/raft.rs]: I think `safe-distance` here is something more like placebo before https://github.com/pingcap/tikv/issues/841 is resolved.

Comment (siddontang) [inline src/raft/raft.rs]: now we can think time is ok. we will consider time problem after @swingbach and us think a solution.

Comment (BusyJay) [inline src/raft/raft.rs]: If time is ok, then `safe-distance` is unnecessary.

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -469,7 +469,31 @@ impl Peer {
             return cmd.cb.call_box((err_resp,));
         }
 
-        if get_transfer_leader_cmd(&req).is_some() {
+        let local_read = req.has_header() && req.get_header().has_read_quorum() &&
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: move this to a function 
if has_read_quorum is true, we should use raft read.

**Comment by siddontang** [comment]
any test? we should test for lease read and quorum read.

**Comment by zhangjinpeng1987** [comment]
@siddontang @ngaut @disksing @BusyJay PTAL

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -513,6 +532,21 @@ impl Peer {
         Ok(())
     }
 
+    fn use_local_read(&self, req: &RaftCmdRequest) -> bool {
+        if !req.has_header() || !req.get_header().has_read_quorum() ||
+           req.get_header().get_read_quorum() || !self.raft_group.raft.in_lease() {
+            return false;
+        }
+
+        for i in 0..req.get_requests().len() {
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: `for cmd_req in req.get_requests()`

File: src/raftstore/store/peer.rs [hunk 5]
```diff
@@ -513,6 +532,21 @@ impl Peer {
...
+            if req.get_requests()[i].get_cmd_type() != CmdType::Snap {
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: I think `CmdType::Get` and `CmdType::Seek` can use local read too.

File: src/raftstore/store/peer.rs [hunk 6]
```diff
@@ -513,6 +532,21 @@ impl Peer {
         Ok(())
     }
 
+    fn use_local_read(&self, req: &RaftCmdRequest) -> bool {
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: IMO, `use_local_read` seems that we will use local read in this function, but this function is just to check whether we need local read or not.

File: src/raftstore/store/peer.rs [hunk 7]
```diff
@@ -513,6 +532,21 @@ impl Peer {
...
+        if !req.has_header() || !req.get_header().has_read_quorum() ||
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: `req.has_header() || !req.get_header().has_read_quorum() || req.get_header().get_read_quorum()` can be collapsed as `req.get_header().get_read_quorum()`.

File: src/raftstore/store/peer.rs [hunk 8]
```diff
@@ -513,6 +532,21 @@ impl Peer {
...
+           req.get_header().get_read_quorum() || !self.raft_group.raft.in_lease() {
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: I think a `!` is missing here.

File: src/storage/engine/raftkv.rs [hunk 9]
```diff
@@ -148,6 +148,9 @@ impl<S: RaftStoreRouter> RaftKv<S> {
         header.set_peer(ctx.get_peer().clone());
         header.set_region_epoch(ctx.get_region_epoch().clone());
         header.set_uuid(Uuid::new_v4().as_bytes().to_vec());
+        if ctx.has_read_quorum() && ctx.get_read_quorum() {
```
Comment (BusyJay) [inline src/storage/engine/raftkv.rs]: `header.set_read_quorum(ctx.get_read_quorum())`

File: src/raftstore/store/peer.rs (hunk 8 above)
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: My mistake. No need `!` here.

**Comment by siddontang** [comment]
You should do more detailed test. E,g if a leader is isolated but still in lease, the no read quorum read can still read data from it, but quorum read will fail.

File: src/raftstore/store/peer.rs [hunk 10]
```diff
@@ -469,7 +469,26 @@ impl Peer {
             return cmd.cb.call_box((err_resp,));
         }
 
-        if get_transfer_leader_cmd(&req).is_some() {
+        let local_read = self.whether_use_local_read(&req);
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: whether seems strange too. @queenypingcap could you help us think a name?

Comment (ngaut) [inline src/raftstore/store/peer.rs]: is_local_read

Comment (queenypingcap) [inline src/raftstore/store/peer.rs]: is_local_read LGTM

**Comment by zhangjinpeng1987** [comment]
PTAL

File: src/raftstore/store/peer.rs (hunk 5 above)
Comment (siddontang) [inline src/raftstore/store/peer.rs]: +1

File: tests/storage/test_raftkv.rs [hunk 11]
```diff
@@ -34,6 +35,54 @@ fn test_raftkv() {
...
+    let storage = cluster.sim.rl().storages[&leader.get_id()].clone();
+
+    let mut ctx = Context::new();
+    ctx.set_region_id(region.get_id());
+    ctx.set_region_epoch(region.get_region_epoch().clone());
+    ctx.set_peer(leader.clone());
+    ctx.set_read_quorum(read_quorum);
+
+    // write some data
+    assert_none(&ctx, storage.as_ref(), b"x");
```
Comment (siddontang) [inline tests/storage/test_raftkv.rs]: use meaningful name like k1, v1 instead of x and 1.

File: tests/storage/test_raftkv.rs [hunk 12]
```diff
@@ -34,6 +35,54 @@ fn test_raftkv() {
...
+    must_put(&ctx, storage.as_ref(), b"x", b"1");
+
+    // isolate leader
+    cluster.add_send_filter(Isolate::new(leader.get_store_id()));
+
+    // leader still in lease, check if can read on leader
+    assert_eq!(can_read(&ctx, storage.as_ref(), b"x"), !read_quorum);
```
Comment (siddontang) [inline tests/storage/test_raftkv.rs]: must check the  read value is last put 1.

**Comment by siddontang** [comment]
@zhangjinpeng1987  CI still fails

File: tests/raftstore/test_multi.rs [hunk 13]
```diff
@@ -37,18 +37,30 @@ fn test_multi_base_after_bootstrap<T: Simulator>(cluster: &mut Cluster<T>) {
     assert_eq!(cluster.get(key), Some(value.to_vec()));
 
     let check_res = cluster.check_quorum(|engine| {
-        match engine.get_value(&keys::data_key(key)).unwrap() {
-            None => false,
-            Some(v) => &*v == value,
+        for _ in 0..50 {
+            match engine.get_value(&keys::data_key(key)).unwrap() {
```
Comment (siddontang) [inline tests/raftstore/test_multi.rs]: can we use must_get instead?

Comment (zhangjinpeng1987) [inline tests/raftstore/test_multi.rs]: I think this test case is useless when CMD_GET use lease read.

**Comment by zhangjinpeng1987** [comment]
@BusyJay @disksing PTAL

**Comment by disksing** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Comment by zhangjinpeng1987** [comment]
@siddontang @BusyJay @ngaut @disksing PTAL

**Comment by zhangjinpeng1987** [comment]
@siddontang @BusyJay @disksing PTAL

File: src/raftstore/store/peer.rs [hunk 14]
```diff
@@ -893,6 +934,7 @@ impl Peer {
         match storage.engine.write(ctx.wb) {
             Ok(_) => {
                 storage.apply_state = ctx.apply_state;
+                storage.applied_index_term = term;
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: L788 also need to be updated.

File: src/raftstore/store/peer_storage.rs [hunk 15]
```diff
@@ -158,6 +159,7 @@ impl PeerStorage {
             snap_sched: snap_sched,
             snap_tried_cnt: AtomicUsize::new(0),
             tag: tag,
+            applied_index_term: 0,
```
Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: The term should be properly initialised when created.

**Comment by siddontang** [comment]
LGTM

File: tests/raftstore/test_multi.rs [hunk 16]
```diff
@@ -346,6 +346,76 @@ fn test_leader_change_with_uncommitted_log<T: Simulator>(cluster: &mut Cluster<T
...
+        .msg_type(MessageType::MsgAppend)
+        .direction(Direction::Recv));
+
+    // Make peer 2 have no way to know the uncommitted entries can be applied
+    // when it's still follower.
+    cluster.add_send_filter(IsolateRegionStore::new(1, 2)
+        .msg_type(MessageType::MsgHeartbeat)
+        .direction(Direction::Recv));
+
+    let (k1, v1) = (b"k1", b"v1");
```
Comment (ngaut) [inline tests/raftstore/test_multi.rs]: s/k1/k/
s/v2/v/

File: tests/raftstore/test_multi.rs [hunk 17]
```diff
@@ -346,6 +346,76 @@ fn test_leader_change_with_uncommitted_log<T: Simulator>(cluster: &mut Cluster<T
...
+    cluster.must_put(k1, v1);
+
+    // peer 1 must have committed, but peer 2 has not.
+    must_get_equal(&cluster.get_engine(1), k1, v1);
+
+    cluster.must_transfer_leader(1, util::new_peer(2, 2));
+
+    // leader's term not equal applied index's term, if we read local, we may get old value
+    // in this situation we need use raft read
```
Comment (ngaut) [inline tests/raftstore/test_multi.rs]: s/need/need to/

**Comment by ngaut** [comment]
LGTM

**Pull request merged**